icicle/src/merkle-tree/mmcs.cu

```diff
@@ -0,0 +1,458 @@
+#include "hash/hash.cuh"
+#include "merkle-tree/merkle.cuh"
+#include "matrix/matrix.cuh"
+
+#include <algorithm>
+
+using matrix::Matrix;
+
+namespace merkle_tree {
+
+  template <typename L, typename D>
+  cudaError_t hash_leaves(
+    const Matrix<L>* leaves,
+    unsigned int number_of_inputs,
+    uint64_t number_of_rows,
+    D* states,
+    D* digests,
+    unsigned int digest_elements,
+    const SpongeHasher<L, D>& hasher,
+    const device_context::DeviceContext& ctx)
+  {
+    SpongeConfig sponge_config = default_sponge_config(ctx);
+    sponge_config.are_inputs_on_device = true;
```

Comment:
what is this assumption about being on device? why not pass the config from outside and avoid this assumption?

Reply:
It's not an assumption, it is a requirement of the current implementation